crates/valence_command/src/parsers.rs

```diff
@@ -0,0 +1,269 @@
+//! A collection of parses for use in command argument nodes.
+pub mod angle;
+pub mod block_pos;
+pub mod bool;
+pub mod color;
+pub mod column_pos;
+pub mod entity_anchor;
+pub mod entity_selector;
+pub mod gamemode;
+pub mod inventory_slot;
+pub mod numbers;
+pub mod rotation;
+pub mod score_holder;
+pub mod strings;
+pub mod swizzle;
+pub mod time;
+pub mod vec2;
+pub mod vec3;
+
+use std::ops::Add;
+
+pub use block_pos::BlockPos;
+pub use column_pos::ColumnPos;
+pub use entity_anchor::EntityAnchor;
+pub use entity_selector::EntitySelector;
+pub use inventory_slot::InventorySlot;
+pub use rotation::Rotation;
+pub use score_holder::ScoreHolder;
+pub use strings::{GreedyString, QuotableString};
+pub use swizzle::Swizzle;
+use thiserror::Error;
+pub use time::Time;
+pub(crate) use valence_server::protocol::packets::play::command_tree_s2c::Parser;
+pub use vec2::Vec2;
+pub use vec3::Vec3;
+
+pub trait CommandArg: Sized {
+    fn arg_from_string(string: String) -> Result<Self, CommandArgParseError> {
+        Self::parse_arg(&mut ParseInput::new(string))
+    }
+
+    fn parse_arg(input: &mut ParseInput) -> Result<Self, CommandArgParseError>;
+    /// what will the client be sent
+    fn display() -> Parser;
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct ParseInput {
+    pub input: String,
+    pub cursor: usize,
+}
+
+impl ParseInput {
```

Comment:
There's some buggy behavior here. A character is not always one byte wide due to UTF-8 encoding. The code below sometimes assumes `cursor` is a count of bytes, while other times it's used as a count of chars. Also, `chars().nth(n)` has to walk the entire length of the string to get to `n`.